### DISCUSSION TITLE
Completely disallow Fully Qualified Class Name (FQCN) references everywhere?

### DISCUSSION THREAD

**Discussion by arderyp** [PHP-CS-Fixer/PHP-CS-Fixer#6966] [Q&A] [answered]
I am using `PhpCsFixer\Fixer\Import\GlobalNamespaceImportFixer`, which converts:

```php
throw new \Exception();
```

to 

```diff
+use Exception;

-throw new \Exception();
+throw new Exception();
```

But I have some FQCN instances coming out of `rector/rector` that aren't getting picked up and fixed.

## Edge Case Number 1
```php
namespace Entity;

class Apple extends \Entity\AbstractEntity
```

The above is not being fixed, but should be converted as follows to remove the unnecessary namespace on `extends` because the parent is in the same namespace as the extending child:
```diff
 namespace Entity;

-class Apple extends \Entity\AbstractEntity
+class Apple extends AbstractEntity
```

## Edge Case Number 2
```php
#[\PHPUnit\Framework\Attributes\DoesNotPerformAssertions]
public function testConstructorSuccess(): void
```

This should be converted as follows:
```diff
+use PHPUnit\Framework\Attributes\DoesNotPerformAssertions;

-#[\PHPUnit\Framework\Attributes\DoesNotPerformAssertions]
+#[DoesNotPerformAssertions]
 public function testConstructorSuccess(): void
```

My fixer config:

```php
<?php declare(strict_types=1);

use PhpCsFixer\Fixer\Import\GlobalNamespaceImportFixer;
use Symplify\EasyCodingStandard\Config\ECSConfig;
use Symplify\EasyCodingStandard\ValueObject\Set\SetList;

return function (ECSConfig $ecsConfig): void {
    $ecsConfig->paths([
        __DIR__ . '/src',
        __DIR__ . '/tests',
    ]);

    $ecsConfig->rules([
        GlobalNamespaceImportFixer::class,  // Force use statement over FQCN reference
    ]);

    $ecsConfig->sets([
        SetList::COMMENTS,
        SetList::NAMESPACES,
        SetList::STRICT,
        SetList::PSR_12,
    ]);
};
```

I'm guessing one of three things is going on:
1. these are bugs/edge cases in `GlobalNamespaceImportFixer`
2. there are some additional Fixers I should use to catch these
3. some new fixers should be created
  * fix unnecessary FQCN references to classes in the same directory
  * fix FQCN references in attribute blocks

I'm happy to debug further id there's any additional info that would help.

**Comment by arderyp** [comment]
@Wirone, thanks for converting this to a discussion.  Do you have any insight on the matter?

**Reply by IgorDePaula** [reply]
Any update?

**Reply by arderyp** [reply]
I never figured out a way to do it 🤷‍♂️

**Comment by Wirone** [comment, marked as answer]
First of all, you asked about PHP CS Fixer's `GlobalNamespaceImportFixer`, but you mention Rector and provide ECS configuration 🤷‍♂️. But anyway...

1. `GlobalNamespaceImportFixer`, as its name speaks explicitly, handles only symbols from global namespace, so it won't fix any FQCN from custom namespace.
2. There is [`fully_qualified_strict_types`](https://cs.symfony.com/doc/rules/import/fully_qualified_strict_types.html) fixer, which on the other hand handles FQCNs, but only for symbols already imported, and not all cases are covered.
3. There are [several issues related to FQCNs](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues?q=is%3Aopen+is%3Aissue+label%3Atopic%2Ffqcn) to improve current support.

PS. [Rector has configuration option to import FQCNs](https://getrector.com/documentation/import-names).

**Reply by arderyp** [reply]
Yes, I leverage ecs to use php-CS-fixer.  I've used rector for other cleanups.  I was aware of most of your points but not the last rector link, thanks :)

**Reply by Wirone** [reply]
There is nothing wrong in using ECS, I've been using it (and contributed to it) for years. But then you probably should ask for help in ECS repo, since debugging ECS rule sets is not Fixer maintainers' responsibility 😉. Fortunately I am ECS and Rector user myself, so I could help.

**Reply by arderyp** [reply]
well i asked here because I'm not having issues with ecs, and ecs doesn't provide the fixers, it just provides and interface to use them.  i showed my ecs config just to demonstrate what fixers i was using.

**Reply by Wirone** [reply]
> ecs doesn't provide the fixers, it just provides and interface to use them

You're wrong. ECS [includes its fixers](https://github.com/easy-coding-standard/easy-coding-standard/blob/e3b4328440c20b681fb6106fef1062154c6236aa/composer.json#L20) from [`symplify/coding-standard`](https://github.com/symplify/coding-standard), also [some rulesets can include sniffs from PHPCodeSniffer](https://github.com/easy-coding-standard/easy-coding-standard/blob/e3b4328440c20b681fb6106fef1062154c6236aa/config/set/common/comments.php#L5), that's why I said it's out of the scope of this repo 😉.

**Reply by arderyp** [reply]
I was simply looking for advice to see if php-CS-fixer provided a fixer for the stated goal, as I know php-CS-fixer has a diverse suite of fixers (that I'm not super knowledgeable about) and which I can leverage via ecs. But if you think this is a radically offensive and unproductive thread, feel free to close it.

**Reply by Wirone** [reply]
I know you wanted help, that's why I responded and provided info from Fixer's point of view. But since ECS has custom fixers and support for sniffs, solution for your problem may be different if responded from ECS point of view 🙂.

If my answer satisfies your need, then you can mark it as accepted.
